Improvements to query syntax

DRK3 opened the pull request:
Made some enhancements to the query syntax:

1. Merged the "name + value" and "has" query types into one query syntax. The "name + value" types of queries allowed for more advanced types of queries involving ANDs and ORs, but you couldn't do a query where you treated the value like a wildcard (the "has" query). With this change, you can mix and match as needed. I removed the "has" query section since it's now redundant.
2. With this new syntax, the format of the query more closely resembles Encrypted Document attributes.
3. Made some improvements to the description of the query syntax to make it easier to understand.

DRK3 commented:
Here are some alternate ideas for the query format and why I didn't go with them (both reasons are edge-cases, but still...):
1. Use a blank value to indicate the wildcard: the problem here is that now there's no way to do a query for name+ blank values specifically without grabbing all name + values with matching names
2. Making it so that when the name is "has", then the right hand side is treated as the name instead of a value: problem here is that technically "has" is a valid base64 string, so you technically create an ambiguous case here

DRK3 commented:
I actually think what might be a better and more flexible solution is to define a query syntax using a single string instead of a nested JSON structure. Something like this... 

A sequence of attribute names or attribute name pairs (name-values separated by colons), with operators like &,| separating them. Could also support adding brackets for controlling the order of expression. These operator characters won't collide with attribute names or values since they're Base64 encoded.

TallTed commented on index.html:
```diff
@@ -2687,11 +2687,15 @@ <h3>
           <p>
   When doing a search for a specific value associated with a specific index, the syntax for the <code>equals</code> filter is as follows:
   <code>equals</code> is an array of one or more subfilters.
```
```suggestion
          </p>
          <ul>
            <li>
  <code>equals</code> is an array of one or more sub-filters.
            </li>
```

TallTed commented on index.html:
```diff
@@ -2687,11 +2687,15 @@ <h3>
-  For a document to match a subfilter, it MUST contain all the given attribute pairs within that subfilter.
-  In other words, key-value attribute pairs within a subfilter indicate an AND operation with those pairs, with the final result being an OR operation between the subfilters.
-  If there is only one subfilter needed, then <code>equals</code> MAY be directly set to it instead of being an array of one element.
+  Each subfilter is an array of one or more criteria.
+
+  A criterion has three possible fields: <code>name</code>, <code>value</code>, and <code>has</code>.
+  A criterion can be used in two different ways. To specify documents that contain a specific name+value pair, set <code>name</code> and <code>value</code> accordingly.
+  To treat the value in a name-value pair as a wildcard, where any value associated with that name will match, use the <code>has</code> keyword instead.
+  When <code>name</code> and <code>value</code> are being used for a single matching criterion, <code>has</code> must be blank. When <code>has</code> is being used, <code>name</code> and <code>value</code> must be blank.
+
+  Multiple criteria in a single subfilter act as AND operations with respect to one another.
+  Subfilters act as OR operations with respect to one another.
```
```suggestion
            <li>
  Each sub-filter is an array of one or more criteria.
            </li>
            <li>
  A criterion has three possible fields: <code>name</code>, <code>value</code>, and <code>has</code>.
            </li>
            <li>
  A criterion can be used in two different ways. 
              <ul>
                <li>
  To specify documents that contain a specific name+value pair, set <code>name</code> and 
  <code>value</code> accordingly.
                </li>
                <li>
  To treat the value in a name-value pair as a wildcard, where any value associated with 
  that name will match, use the <code>has</code> keyword instead.
                </li>
              </ul>
            </li>
            <li>
  When <code>name</code> and <code>value</code> are being used for a single matching criterion, 
  <code>has</code> must be blank. When <code>has</code> is being used, <code>name</code> and 
  <code>value</code> must be blank.
            </li>
            <li>
  Multiple criteria in a single sub-filter act as <code>AND</code> operations with respect 
  to one another.
            </li>
            <li>
  Sub-filters act as <code>OR</code> operations with respect to one another.
            </li>
          </ul>
```

TallTed reviewed (changes requested):
This (or similar) bulletization was clearly intended by the text in the HTML source, but would not be rendered as such without the HTML elements I've suggested.

DRK3 commented:
@TallTed Great suggestion - I've added list tags and it looks much better now. Thanks!

I also made some other improvements:
1. I've italicized key terms like "subfilter" and "criterion" in an effort to make it more clear that they refer to specific pieces of the overall syntax
2. I've updated the example titles to use these terms explicitly
3. I also updated the final query example to incorporate a "has" query along with "name-value" queries to show a "full mix" of all types of operations within a single query.
